vm/systemSmartContracts/governance.go

```diff
@@ -147,73 +165,268 @@ func (g *governanceContract) init(args *vmcommon.ContractCallInput) vmcommon.Ret
+	if err != nil {
+		g.eei.AddReturnMessage(err.Error())
 		return vmcommon.UserError
 	}
-	minVeto, okConvert := big.NewInt(0).SetString(string(args.Arguments[2]), conversionBase)
-	if !okConvert || minVeto.Cmp(big.NewInt(0)) < 0 {
-		g.eei.AddReturnMessage("changeConfig third argument is incorrectly formatted")
+
+	err = g.saveNewVoteData(voterAddress, newVoteData, updatedProposal, Account)
+	if err != nil {
+		g.eei.AddReturnMessage(err.Error())
+		return vmcommon.ExecutionFailed
+	}
+
+	return vmcommon.Ok
+}
+
+// validatorVote casts a vote for a validator. This function can receive 3 or 4 parameters:
+//  args.Arguments[0] - proposal reference (github commit)
+//  args.Arguments[1] - vote option (yes, no, veto)
+//  args.Arguments[2] - vote power used for this vote
+//  args.Arguments[3] (optional) - an address that identifies if the vote was made on behalf of someone else - this
+//   only helps for statistical and view purposes, it does not afftect the logic in any way
+func (g *governanceContract) validatorVote(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	err := g.eei.UseGas(g.gasCost.MetaChainSystemSCsCost.Vote)
+	if err != nil {
+		g.eei.AddReturnMessage("not enough gas")
+		return vmcommon.OutOfGas
+	}
+
+	if len(args.Arguments) < 3 || len(args.Arguments) > 4 {
+		g.eei.AddReturnMessage("invalid number of arguments, expected 3 or 4")
+		return vmcommon.FunctionWrongSignature
+	}
+
+	voterAddress := args.CallerAddr
+	proposalToVote := args.Arguments[0]
+	proposal, err := g.getValidProposal(proposalToVote)
+	if err != nil {
+		g.eei.AddReturnMessage(err.Error())
 		return vmcommon.UserError
 	}
-	minPass, okConvert := big.NewInt(0).SetString(string(args.Arguments[3]), conversionBase)
-	if !okConvert || minPass.Cmp(big.NewInt(0)) < 0 {
-		g.eei.AddReturnMessage("changeConfig fourth argument is incorrectly formatted")
+	voteOption, err := g.castVoteType(string(args.Arguments[1]))
+	if err != nil {
+		g.eei.AddReturnMessage(err.Error())
 		return vmcommon.UserError
 	}
 
-	scConfig, err := g.getConfig()
+	votePower := big.NewInt(0).SetBytes(args.Arguments[2])
+	delegatedTo, err := g.getDelegatedToAddress(args)
 	if err != nil {
-		g.eei.AddReturnMessage("changeConfig error " + err.Error())
+		g.eei.AddReturnMessage(err.Error())
+		return vmcommon.UserError
+	}
+	totalVotingPower, err := g.getValidatorVotingPower(voterAddress)
+	if err != nil {
+		g.eei.AddReturnMessage(err.Error())
 		return vmcommon.UserError
 	}
 
-	scConfig.NumNodes = numNodes.Int64()
-	scConfig.MinQuorum = int32(minQuorum.Int64())
-	scConfig.MinVetoThreshold = int32(minVeto.Int64())
-	scConfig.MinPassThreshold = int32(minPass.Int64())
+	currentVoteData, err := g.getOrCreateVoteData(proposalToVote, voterAddress)
+	if err != nil {
+		g.eei.AddReturnMessage(err.Error())
+		return vmcommon.ExecutionFailed
+	}
+	if totalVotingPower.Cmp(big.NewInt(0).Add(votePower, currentVoteData.UsedPower)) == -1 {
```

Comment:
I think it's a check not to vote more than the power: 
if totalVotingPower < votePower + usedPower,
but what is this check used for?

Reply:
Like discussed, the bug was in the flow where a normal account votes multiple times. In that case the voting power needs to be adjusted so that the sum of the voting power used in multiple small votes would be sqrt(totalFundsUsed). Fixed

Reply:
ok